src/promise.h

```
@@ -1,28 +1,96 @@
 #pragma once
 
 #include <kj/async.h>
+#include <kj/debug.h>
 #include <rust/cxx.h>
 
+// These types are shared with rust
 namespace kj_rs {
 
 using OwnPromiseNode = kj::_::OwnPromiseNode;
 
-void own_promise_node_drop_in_place(OwnPromiseNode*);
-
-// https://github.com/dtolnay/cxx/blob/86cd652c06c5cb4c2e24d3ab555cf707b4ae0883/src/cxx.cc#L518
 namespace repr {
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wreturn-type-c-linkage"
+
+// ::cxx::private::PtrLen
 struct PtrLen final {
-  void *ptr;
-  std::size_t len;
+  void *ptr = nullptr;
+  std::size_t len = 0;
+};
+
+extern "C" {
+repr::PtrLen cxxbridge1$exception(const char *, std::size_t len) noexcept;
+}
+
+// ::cxx::private::Result
+struct Result final {
+  PtrLen err = {};
+  inline static Result ok() { return {}; }
+  inline static Result error(kj::Exception &e);
 };
+
+// ::kj::promise::UnwrapCallback
+using UnwrapCallback = Result (*)(const void /* kj::_::PromiseNode */ *node,
+                                  void /* T */ *ret);
+// ::kj::promise::KjPromiseNodeImpl
+struct KjPromiseNodeImpl {
+  template <typename T> inline KjPromiseNodeImpl(kj::Promise<T> &&p);
+
+  kj::_::PromiseNode *node;
+  repr::UnwrapCallback unwrap = nullptr;
```

Nit: Our one constructor initializes both fields, so this non-static default initializer looks a bit misleading. If we ever rely on it, we would need an NSDI for `node`, too.